contracts/src/ERC20Destination.sol

```diff
@@ -96,6 +110,39 @@ contract ERC20Destination is IERC20Bridge, TeleporterTokenDestination, ERC20 {
         _burn(address(this), amount);
     }
 
+    function _handleSendAndCall(
+        SingleHopCallMessage memory message,
+        uint256 amount
+    ) internal virtual override {
+        // Mint the tokens to this contract address.
+        _mint(address(this), amount);
+
+        // Approve the destination contract to spend the amount.
+        _approve(address(this), message.recipientContract, amount);
+
+        // Encode the call to {IERC20SendAndCallReceiver-receiveTokens}
+        bytes memory payload = abi.encodeCall(
+            IERC20SendAndCallReceiver.receiveTokens,
+            (address(this), amount, message.recipientPayload)
+        );
+
+        // Call the destination contract with the given payload and gas amount.
+        bool success = GasUtils._callWithExactGas(
+            message.recipientGasLimit, message.recipientContract, payload
+        );
+
+        // Reset the destination contract allowance to 0.
+        _approve(address(this), message.recipientContract, 0);
```

Comment:
I added this behavior for `ERC20Source` and `ERC20Destination`. Unfortunately it's not possible for the native source/destination contracts because contracts are credits the full `msg.value` amount prior to when a `payable` function is called. I added comments noting that `INativeTokenReceiver` implementation are responsible for ensuring that those coins are handled properly.